File: VFIT_Backend/src/main/java/com/vfit/modules/nutrition/entity/Food.java
package com.vfit.modules.nutrition.entity;

import com.fasterxml.jackson.annotation.JsonAlias;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Field;
import org.springframework.data.mongodb.core.mapping.Document;

@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "foods")
@CompoundIndex(
        name = "idx_food_gym_friendly_popularity",
        def = "{'is_gym_friendly': -1, 'popularity_score': -1}")
public class Food {
    @Id
    private String id;

    @Indexed(name = "idx_food_name")
    private String name;

    @Field("normalized_name")
    @Indexed(name = "idx_food_normalized_name")
    private String normalizedName;

    @Field("serving_size_grams")
    private double servingSizeGrams;

    private double calories;
    private double protein;
    private double carbs;
    private double fat;

    @Field("calories_per_100g")
    private double caloriesPer100g;
    @Field("protein_per_100g")
    private double proteinPer100g;
    @Field("carbs_per_100g")
    private double carbsPer100g;
    @Field("fat_per_100g")
    private double fatPer100g;
    @Field("is_gym_friendly")
    @JsonAlias({"isGymFriendly", "is_gym_friendly"})
    private boolean isGymFriendly;
    @Field("search_count")
    private int searchCount;
    @Field("popularity_score")
    private int popularityScore;
}

File: VFIT_Fontend/lib/features/nutrition/domain/entities/food.dart
import 'package:equatable/equatable.dart';

class Food extends Equatable {
  const Food({
    required this.id,
    required this.name,
    required this.normalizedName,
    required this.servingSizeGrams,
    required this.calories,
    required this.protein,
    required this.carbs,
    required this.fat,
    required this.caloriesPer100g,
    required this.proteinPer100g,
    required this.carbsPer100g,
    required this.fatPer100g,
    required this.isGymFriendly,
    required this.searchCount,
    required this.popularityScore,
  });

  final String id;
  final String name;
  final String normalizedName;
  final double servingSizeGrams;
  final double calories;
  final double protein;
  final double carbs;
  final double fat;
  final double caloriesPer100g;
  final double proteinPer100g;
  final double carbsPer100g;
  final double fatPer100g;
  final bool isGymFriendly;
  final int searchCount;
  final int popularityScore;

  double get _servingDivisor => servingSizeGrams <= 0 ? 100 : servingSizeGrams;

  double caloriesFor(double grams) => grams / _servingDivisor * calories;
  double proteinFor(double grams) => grams / _servingDivisor * protein;
  double carbsFor(double grams) => grams / _servingDivisor * carbs;
  double fatFor(double grams) => grams / _servingDivisor * fat;

  @override
  List<Object?> get props => [
        id,
        name,
        normalizedName,
        servingSizeGrams,
        calories,
        protein,
        carbs,
        fat,
        caloriesPer100g,
        proteinPer100g,
        carbsPer100g,
        fatPer100g,
        isGymFriendly,
        searchCount,
        popularityScore,
      ];
}
